src/ui/edit/machine-canvas-item.c

```diff
@@ -194,6 +200,58 @@ update_machine_graphics (BtMachineCanvasItem * self)
   g_object_unref (pixbuf);
 }
 
+static void
+update_custom_graphics (BtMachineCanvasItem * self, guint width, guint height, guint32 * data) {
+  g_return_if_fail(self->priv->image_custom_gfx);
+
+  if (data) {
+    clutter_image_set_data (CLUTTER_IMAGE (self->priv->image_custom_gfx),
+        (guint8*)data, COGL_PIXEL_FORMAT_RGBA_8888, width, height, width*4, NULL);
+  } else {
+    int32_t zero = 0;
+    clutter_image_set_data (CLUTTER_IMAGE (self->priv->image_custom_gfx),
+        (guint8*)&zero, COGL_PIXEL_FORMAT_RGBA_8888, 1, 1, 4, NULL);
+  }
+}
+
+static gboolean
+custom_gfx_request_after_invalidation (BtMachineCanvasItem * self) {
+  GstElement *element;
+  g_object_get (self->priv->machine, "machine", &element, NULL);
+  
+  const BtUiCustomGfx* gfx = 0;
+  g_signal_emit_by_name (element, "bt-gfx-request", &gfx);
+  if (gfx) {
+    update_custom_graphics (self, gfx->width, gfx->height, gfx->data);
+  } else {
+    update_custom_graphics (self, 0, 0, NULL);
+  }
+  
+  g_clear_object (&element);
+          
+  g_mutex_lock (&self->priv->custom_gfx_lock);
+  self->priv->custom_gfx_timer = 0;
+  g_mutex_unlock (&self->priv->custom_gfx_lock);
+  return FALSE;
+}
+
+static void
+on_custom_gfx_invalidated (BtMachine * machine, BtMachineCanvasItem * self) {
+  // Updating the custom gfx this way is important for two reasons:
+  //
+  // 1. It makes sure that the image update happens on the main thread.
+  //    Clutter doesn't seem to like requests coming from another thread.
+  //    This will likely be called from the Gstreaming processing thread.
+  // 2. It aggregates many invalidating requests (i.e. from multiple machine
+  //    property updates) into a single "update gfx" call.
+  g_mutex_lock(&self->priv->custom_gfx_lock);
+  if (!self->priv->custom_gfx_timer)
+    self->priv->custom_gfx_timer =
+        g_timeout_add(5, G_SOURCE_FUNC(custom_gfx_request_after_invalidation),
+            self);
```

Comment:
I think you can use g_idle_add for this

Reply:
Ok, great!

I didn't even do my maths properly here and so didn't set the timeout value properly, but my intention here was to limit the gfx refreshes to roughly 60fps to avoid wasting CPU. The timeout value should be 16, not 5.

Do you think that using the idle function will use more CPU in practice?

Reply:
Actually, you are right. A timer makes more sense here. You will periodically request the machine to update the image buffer.

Reply:
Not quite -- the update aren't periodic. The graphics are only flagged for an update when the machine's "bt-gfx-invalidated" signal is emitted. Currently this happens when the machine is first assigned to the BtMachineCanvasItem, and then whenever the machine itself wants to request an update.

I imagine the machine will almost always be emitting "bt-gfx-invalidated" when one of its properties is updated. This is what my current implementation is doing. However, a machine could also request updates periodically if it wants to do some "realtime" display. So there's some flexibility there.

You see bursts of updates when the song starts, and also when the user is dragging sliders back and forth. So I was thinking this is the reason for the timer over the idle callback -- I think the idle callback could easily end up making more than 60fps of graphics updates during these bursts, even though it wasn't updating on every call. However, the timer would limit the max update rate. But I might have the wrong idea about how often/quickly the idle callback gets called.

Reply:
Fixed that timer value and squashed commits, btw.

Reply:
The bursts of updates are because when you set the song to play, the CPU quickly spikes to flood the pipeline with buffers (preroll). Once all elements in a pipeline have buffers, the audio-sinks and the queue's stall upstream branches to only send more data as it is needed.
I did not thought of that when doing the review (have been away from the code for too long). Anyway please take a look at e.g. gstreamer Level element (or Spectrum) and how it is used in buzztrax. The data that is emitted has the play-back time and I use a g_timeout_add to defer drawing to the point in time when the audio is played. This is especially noticable for sources with lots of fx.
Maybe we can do the same? Emit the data as a GstMessage on the bus and include a timestamp? Another option is an GstEvent that is emitted by the sink as a GstMessage (https://gstreamer.freedesktop.org/documentation/gstreamer/gstevent.html?gi-language=c#gst_event_new_sink_message). This has the advantage that Gsteamer will handle the syncronisation, but there is a little overhead on passing the data along and if you have multiple sinks (e.g. when rendering a wav to disk) the data is copied and the app needs to deduplicate the Message.

Finally, regarding he interface - I think we don't need to worry right now (if the GstMessage works), but the interface can be optional. The element would still work as a normal GstElement.

Reply:
Thank you, I'll take a look at Spectrum to see how it works. One difference I can immediately think of is that unlike spectrum analysis, the generation of the sound isn't usually the "event" leading to the update. In my mind the most common trigger is a UI interaction with a parameter, and the next most common a pattern position change. This applies to the kinds of visualisations I'm using now, which are just parameter-based. 

Although, there would be other classes of visualisations that would be more like the spectrum case, i.e. if an author wanted to visualise k-rate or a-rate internal values of the machine (I'm using some CSound terms here,) so it might might sense to collapse everything to this case.

I'm very happy to let this one bake and iterate on it until we have a good approach to work with.